### ISSUE TITLE
[question] How to add compiler sanitisers to conan

### ISSUE THREAD

**Issue by ghost** [conan-io/conan#15374]
### What is your question?

Hey guys,
I'm invoking conan from Cmake. We are still using conan 1.x since to conan 2.0 cmake wrapper seams to be still a work in progress.

According to this page there are several ways to pass the sanitiser flags to conan.
https://docs.conan.io/1/howtos/sanitizers.html
I decided to go with a custom profile. We are using Ninja MultiConfig with MSVC 17. So we actually have 4 profiles (Debug, Rel, RelWithDeb, MinSizeRel) and we are invoking conan for all 4 configurations (via conan_cmake_configure() and conan_cmake_install())

My profile looks like this

```
[settings]
os=Windows
arch=x86_64
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=17
build_type=Debug
compiler.cppstd=20


[conf]
tools.build:cxxflags=["/fsanitize=address"]
tools.build:cflags=["/fsanitize=address"]
```
But when invoking conan, the compiler output tells me that I'm using ASAN but without /DEBUG. Does build_type=Debug not imply a /DEBUG flag being appended to the c/cxx Fags

How would you modify the profile that ASAN works ?. Thx for your help guys :)

BTW Happy new year :)

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by ghost** [comment]
hey guys,
I did some digging. It looks like tools.build:cxxflags **APPENDS** the sanitiser flags. But some MSVC default flags interfer with the ASAN. Is there any way how to **REPLACE** the compiler flags?

**Comment by memsharded** [comment]
> I'm invoking conan from Cmake. We are still using conan 1.x since to conan 2.0 cmake wrapper seams to be still a work in progress.

Not really. The default branch in the repo has been changed to ``develop2`` already, and the integration is already being used by the CLion plugin and other users quite successfully. It is not less "stable" than the previous 1.X that was not stable either, so it is very recommended to move to Conan 2.0 as soon as possible

**Comment by memsharded** [comment]
> I did some digging. It looks like tools.build:cxxflags APPENDS the sanitiser flags. But some MSVC default flags interfer with the ASAN. Is there any way how to REPLACE the compiler flags?

Conan has the syntax ``+=`` and ``=+`` to prepend and append configuration items as list. The problem is that this is not coming from an appending on compiler flags, but CMake default internal behavior, this is not really Conan, but CMake itself is deciding which flags it passes for a given build-type, if it passes ``/DEBUG`` or not.

You can provide custom toolchain files to configure CMake via ``conf`` ``tools.cmake.cmaketoolchain:user_toolchain``, with that you might inhibit or change some of CMake default flags, maybe that helps.

**Comment by ghost** [comment]
@memsharded  thx for your answer. I'll try to create a custom toolchain for MSVC. 
Not sure if I can get it to work though, I'm used for MSVC to "work out of the box" and do all the setup for me :D

**Comment by memsharded** [comment]
> Not sure if I can get it to work though, I'm used for MSVC to "work out of the box" and do all the setup for me :D

If passing a ``user_toolchain`` doesn't fit well your flow, maybe just editing the ``conan_provider.cmake`` code you can inject what you need there? 

In any case, please keep us posted and don't hesitate to create new tickets as necessary for any further question or issue you might have. Thanks! :)

**Comment by ghost** [comment]
thx @memsharded  for your answer.
After the next release is done, I'll focus on this issue again.
But meanwhile could you tell me what the conan_provider.cmake is.
I could find anything in the docs (https://docs.conan.io/1/search.html?q=conan_provider&check_keywords=yes&area=default)

Is this a generated file?

**Comment by memsharded** [comment]
Closed by https://docs.conan.io/2/security/sanitizers.html, please open new tickets for further questions, thanks for the feedback.
